libfoundation/src/system-library.cpp

```diff
@@ -0,0 +1,302 @@
+/*                                                                     -*-c++-*-
+Copyright (C) 2017 LiveCode Ltd.
+
+This file is part of LiveCode.
+
+LiveCode is free software; you can redistribute it and/or modify it under
+the terms of the GNU General Public License v3 as published by the Free
+Software Foundation.
+
+LiveCode is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or
+FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+for more details.
+
+You should have received a copy of the GNU General Public License
+along with LiveCode.  If not see <http://www.gnu.org/licenses/>.  */
+
+#include "system-private.h"
+
+#include <foundation-auto.h>
+
+/* ---------------------------------------------------------------- */
+
+/* The MCSLibraryRef custom typeinfo. */
+MCTypeInfoRef kMCSLibraryTypeInfo = nullptr;
+
+/* The MCSLibraryCouldNotLoadError error. */
+MCErrorRef kMCSLibraryCouldNotLoadError = nullptr;
+
+/* ================================================================
+ * Types
+ * ================================================================ */
+
+#if defined(__WINDOWS__)
+#include "system-library-w32.hpp"
+#elif defined(__MAC__)
+#include "system-library-mac.hpp"
+#elif defined(__IOS__)
+#include "system-library-ios.hpp"
+#elif defined(__LINUX__)
+#include "system-library-linux.hpp"
+#elif defined(__ANDROID__)
+#include "system-library-android.hpp"
+#endif
```

Comment:
Hmm - the way I've done it at the moment means that this source file needs the __MCSLibraryHandle type defined within it via an inclusion it would be tricky to change that the way things are currently structured. Really it should be a virtual interface I suppose...